test/e2e/operator/deployment_api.go

```diff
@@ -690,7 +722,7 @@ var _ = deploy.DescribeForSome("API", func(d *deploy.Deployment) bool {
 				},
 				"metrics service": func(dep *api.Deployment) apiruntime.Object {
 					obj := &corev1.Service{}
-					key := runtime.ObjectKey{Name: dep.MetricsServiceName(), Namespace: d.Namespace}
+					key := runtime.ObjectKey{Name: base.MetricsServiceName(dep), Namespace: d.Namespace}
```

Comment:
Can this code churn be avoided?

Reply:
Yes, this can be avoided. There's no reason for moving these helper functions into the base package.

Partial patch:
```
diff --git a/pkg/apis/pmemcsi/base/base_types.go b/pkg/apis/pmemcsi/base/base_types.go
index 770600bcb..7bd6b2c11 100644
--- a/pkg/apis/pmemcsi/base/base_types.go
+++ b/pkg/apis/pmemcsi/base/base_types.go
@@ -349,12 +349,6 @@ func ControllerServiceName(d metav1.Object) string {
 	return GetHyphenedName(d) + "-controller"
 }
 
-// MetricsServiceName returns the name of the controller metrics
-// Service object used by the deployment
-func MetricsServiceName(d metav1.Object) string {
-	return GetHyphenedName(d) + "-metrics"
-}
-
 // ServiceAccountName returns the name of the ServiceAccount
 // object used by the deployment
 func ServiceAccountName(d metav1.Object) string {
diff --git a/pkg/apis/pmemcsi/v1beta1/deployment_types.go b/pkg/apis/pmemcsi/v1beta1/deployment_types.go
index 8cf85adb3..68c8b6eb9 100644
--- a/pkg/apis/pmemcsi/v1beta1/deployment_types.go
+++ b/pkg/apis/pmemcsi/v1beta1/deployment_types.go
@@ -158,3 +158,10 @@ func (d *Deployment) GetOwnerReference() metav1.OwnerReference {
 func (d *Deployment) HaveCertificatesConfigured() (bool, error) {
 	return d.Spec.HaveCertificatesConfigured()
 }
+
+// MetricsServiceName returns the name of the controller metrics
+// Service object used by the deployment
+func (d *Deployment) MetricsServiceName() string {
+	// TODO: also move GetHyphenedName and all other functions
+	return base.GetHyphenedName(d) + "-metrics"
+}
diff --git a/pkg/pmem-csi-operator/controller/deployment/controller_driver.go b/pkg/pmem-csi-operator/controller/deployment/controller_driver.go
index 346a5e3dc..3f13403e2 100644
--- a/pkg/pmem-csi-operator/controller/deployment/controller_driver.go
+++ b/pkg/pmem-csi-operator/controller/deployment/controller_driver.go
@@ -497,7 +497,7 @@ func (d *PmemCSIDriver) redeployMetricsService(r *ReconcileDeployment) (apirunti
 		object: func() apiruntime.Object {
 			return &corev1.Service{
 				TypeMeta:   metav1.TypeMeta{Kind: "Service", APIVersion: "v1"},
-				ObjectMeta: d.getObjectMeta(base.MetricsServiceName(d), false),
+				ObjectMeta: d.getObjectMeta(d.MetricsServiceName(), false),
 			}
 		},
 		modify: func(o apiruntime.Object) error {
@@ -653,7 +653,7 @@ func (d *PmemCSIDriver) HandleEvent(meta metav1.Object, obj apiruntime.Object, r
 		klog.Infof("Redeploying service: %s", meta.GetName())
 		if meta.GetName() == base.ControllerServiceName(d) {
 			_, err = d.redeployControllerService(r)
-		} else if meta.GetName() == base.MetricsServiceName(d) {
+		} else if meta.GetName() == d.MetricsServiceName() {
 			_, err = d.redeployMetricsService(r)
 		}
 		if err != nil {
diff --git a/test/e2e/operator/deployment_api.go b/test/e2e/operator/deployment_api.go
index 297707f63..8b32bcac1 100644
--- a/test/e2e/operator/deployment_api.go
+++ b/test/e2e/operator/deployment_api.go
@@ -620,7 +620,7 @@ var _ = deploy.DescribeForSome("API", func(d *deploy.Deployment) bool {
 				},
 				"metrics service": func(dep *api.Deployment) apiruntime.Object {
 					return &corev1.Service{
-						ObjectMeta: metav1.ObjectMeta{Name: base.MetricsServiceName(dep), Namespace: d.Namespace},
+						ObjectMeta: metav1.ObjectMeta{Name: dep.MetricsServiceName(), Namespace: d.Namespace},
 					}
 				},
 				"provisioner role": func(dep *api.Deployment) apiruntime.Object {
@@ -722,7 +722,7 @@ var _ = deploy.DescribeForSome("API", func(d *deploy.Deployment) bool {
 				},
 				"metrics service": func(dep *api.Deployment) apiruntime.Object {
 					obj := &corev1.Service{}
-					key := runtime.ObjectKey{Name: base.MetricsServiceName(dep), Namespace: d.Namespace}
+					key := runtime.ObjectKey{Name: dep.MetricsServiceName(), Namespace: d.Namespace}
 					EventuallyWithOffset(1, func() error {
 						return client.Get(context.TODO(), key, obj)
 					}, "2m", "1s").ShouldNot(HaveOccurred(), "get metrics service set")
```

Reply:
> Can this code churn be avoided?

My initial changes were as you proposed. But, I felt like code duplication which we could avoid. I mean all(`alpha`, `beta`, and future) versions need not define the same function. So I moved them to the API `base` package.

My definition of base package is the 'common' code that could be shared by all versions without duplicating the code.

Reply:
> I mean all(alpha, beta, and future) versions need not define the same function.

Why? It's only called for the API revision that the operator code uses, so the function is only needed there. Or am I missing something?

The cleanest solution would be to not define those helper functions in the API package at all. They are not part of the *API*, they just support code in the operator which works with the API object.